README
mysql-benchmark

The available tools for benchmarking MySQL databases are either too unflexible
or too buggy. So I wrote a new one.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc MySQL::Benchmark

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=mysql-benchmark

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/mysql-benchmark

    CPAN Ratings
        http://cpanratings.perl.org/d/mysql-benchmark

    Search CPAN
        http://search.cpan.org/dist/mysql-benchmark/


LICENSE AND COPYRIGHT

Copyright (C) 2013 Luis Motta Campos

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 dated June, 1991 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.



Changes
Revision history for mysql-benchmark

1.00   Tue 2 Jul 2013 15:53:26 +0200
        Initial release of MySQL::Benchmark

1.01   Tue 2 Jul 2013 18:30:00 +0200
        Implemented a better output method for MySQL::Benchmark.

1.02   Thu 4 Jul 2013 16:30:00 +0200
        Improved query distribution (better approximation to the Normal Curve).
        Added Query Weight feature to better mimic a given workload.

1.03   Fri 5 Jul 2013 16:30:00 +0200
        Added CSV file creation features.
        Added report-to-file feature.
        Improved logging, debugging features.

1.04   Fri 5 Jul 2013 17:10:00 +0200
        Bug fix - was using the wrong variable when closing outfile.
		Added a "quiet" logging option.


lib/MySQL/Benchmark/Query.pm
package MySQL::Benchmark::Query;
use 5.006;
use strict;
use warnings FATAL => 'all';

=head1 NAME

MySQL::Benchmark::Query - an MySQL::Benchmark randomized query abstraction

=head1 SYNOPSIS

    my $query = MySQL::Benchmark::Query->new($yaml_query);
    my $sth   = eval {
        # Returns just the parametrized SQL code.
        $dbh->prepare( $query->sql );
        # Returns randomized parameters accordingly to the YAML description for every call.
        $sth->execute( $query->parameters );
    };
    die if $@;

=head1 DESCRIPTION

=head1 QUERY DESCRIPTION

The Query Description is a data-structure (usually generated from a
configuration file) containing three main elements: an ID String, the
parametrised SQL code to execute (with usual SQL-style placeholders), and an
array of rules to generate arguments.

=head2 ID String (key C<id>)

It's just a string. By convention we'll use the Query-ID generated by
L<pt-query-digest>, which is supposed to be unique and is easy to handle. This
will be the identifier key used on the benchmark reports.

=head2 Query Weight (key C<weight>)

The query weight is just an integer. The higher the integer the more a given
query will appear in the workload. The exact proportion will be calculated by
adding up all Query Weight values in the list of queries, and then normalising
it to a value between 1 and 100, in such a way that every query has at least
weight 1 and in total the sum of all normalised weights is 100.

=head2 Parametrised SQL Code (key C<sql>)

Just plain old SQL with placeholders. Nothing special here.

=head2 The Arguments Generation Array (key C<arguments>)

This is where the magic takes place. This array must contain a list of hashes,
every one of them containing the key C<type> and several other keys depending
on the type. This type and keys will be used to put together randomised
arugments for the queries, that can be returned in order or obtained as an
ordered array of subroutines.

Below we'll desribe each key and it's mandatory parameters.

=head2 Argument Types

Type tells the Query class what is expected from this SQL parameter. Available
types are:

=head3 C<constant>

The C<constant> type evaluates to the same parameter every time. This is the
simplest (and probably the least useful) type. When specifying the C<constant>
type, please provide:

=over 4

=item C<value>

The desired value for the constant.

=item C<quantity> (optional, default: 1)

How many values every call will produce (returned as a list).

=back

Example:

    {   id        => 'Example Query with Constant Parameter',
        weight    => 35,
        sql       => 'SELECT id FROM some_table WHERE idx_val = ?',
        arguments => [ { type => 'constant', value => 10 } ]
    }

=head3 C<list>

The C<list> type evaluates to the list provided every time.

=over 4 

=item C<values> 

The ordered list of values to be returned by this parameter.

=back

Example:

    {   id        => 'Example Query with List Parameter',
        weight    => 35,
        sql       => 'SELECT * FROM some_table WHERE idx_val IN ( ?, ?, ?, ?, ? )',
        arguments => [ { type => 'list', values => [ 1, 2, 3, 4, 5 ] } ]
    }

=head3 C<random_integer>

The C<random_integer> type evaluates to a random value from a range provided.
Additional parameters are:

=over 4

=item C<min>

The floor integer in the range (inclusive).

=item C<max>

The ceiling integer in the range (inclusive).

=item C<quantity> (optional. default: 1)

How many values every call will produce, returned as a list.

=back

Example:

    {   id        => 'Random Integer Example Query',
        weight    => 35,
        sql       => 'SELECT * FROM some_table WHERE idx_val = ?',
        arguments => [ { type => 'random_integer', min => 1, max => 10 } ]
    }

=head3 C<random_from_list>

The C<random_from_list> type evaluates to a random value choosen from a list of
elements. Additional parameters are:

=over 4 

=item C<values>

A list of values from where to choose randomly.

=item C<quantity> (optional. default: 1)

How many items each call will produce, returned as a list.

=back

Example:

    {   id        => 'Random List Pick Example',
        weight    => 35,
        sql       => 'SELECT * FROM some_table WHERE a_column = ?',
        arguments => {
            type       => 'random_from_list',
            values => [ 'A', 'B', 'C', 'D', 'E' ]
        },
    }

=head1 METHODS

=head2 new 

Constructor.

=cut 

sub new {
    my ( $class, $query_descriptor ) = @_;
    my $self = bless {}, $class;
    $self->__process_query_descriptor($query_descriptor);
    return $self;
}

{
    my %function = (
        constant => sub {
            my $arguments = shift;
            my ( $value, $quantity ) = @{$arguments}{qw( value quantity )};
            $quantity ||= 1;
            return sub {
                my @list;
                push @list, $value while $quantity-- > 0;
                return @list;
                }
        },
        list => sub {
            my $arguments = shift;
            my $values    = $$arguments{values};
            return sub { return @$values };
        },
        random_integer => sub {
            my $arguments = shift;
            my ( $min, $max, $quantity )
                = @{$arguments}{qw(min max quantity)};
            $quantity ||= 1;
            return sub {
                my @list;
                push @list, int( rand($max) ) + $min while $quantity-- > 0;
                return @list;
            };
        },
        random_from_list => sub {
            my $arguments = shift;
            my ( $list, $quantity ) = @{$arguments}{qw(list quantity)};
            $quantity ||= 1;
            return sub {
                my @list;
                push @list, $$list[ int( rand( scalar(@$list) ) ) ]
                    while $quantity-- > 0;
                return @list;
            };
        },
    );

    sub __process_query_descriptor {
        my ( $self, $query_descriptor ) = @_;

        die 'No query ID?!' unless exists $$query_descriptor{id};
        $$self{id} = $$query_descriptor{id};

        $$self{weight} = $$query_descriptor{weight};
        $$self{weight} ||= 1; # Default

        die 'No SQL?!' unless exists $$query_descriptor{sql};
        $$self{sql} = $$query_descriptor{sql};

        die 'No arguments?!' unless exists $$query_descriptor{arguments};
        $$self{__parameters} = [
            map {
                die qq(There's no generation function for type '$$_{type}'.)
                    unless exists $function{ $$_{type} };
                $function{ $$_{type} }->($_)
            } @{ $$query_descriptor{arguments} }
        ];
    }
}

=head2 id

Accessor. Returns the ID for this query.

=cut 

sub id { ${ $_[0] }{id} }

=head2 weight

Accessor. Returns the weight associated with this query.

If a second value is provided, it will be stored as this instance's weight.

=cut 

sub weight { defined $_[1]? ${ $_[0] }{weight} = $_[1] : ${ $_[0] }{weight} }

=head2 sql

Accessor. Returns the Parametrised SQL Code represented by this query object,
suitable for the L<DBI::prepare> method call.

=cut

sub sql { ${ $_[0] }{sql} }

=head2 parameters

Returns a list of parameters suitable to be fed to L<DBI::execute>, containing
as many arguments as specified and in the same order as the specification in
the C<arguments> input parameter.

=cut 

sub parameters {
    my ($self) = @_;
    return map { $_->() } @{ $$self{__parameters} };
}

=head1 AUTHOR

Luis Motta Campos, C<< <lmc at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Luis Motta Campos.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; version 2 dated June, 1991 or at your option any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree; if
not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.

=cut

1;    # End of MySQL::Benchmark::Query



lib/MySQL/Benchmark/Logger.pm
package MySQL::Benchmark::Logger;

use 5.006;
use warnings FATAL => 'all';
use strict;

use Time::HiRes ();
use FileHandle;
use base qw( Exporter );
our @EXPORT_OK = qw( log );

BEGIN { $_->autoflush(1) for \*STDOUT, \*STDERR }

=head1 NAME

MySQL::Benchmark::Logger - Logging facility wrapper for MySQL::Benchmark

=head1 SYNOPSYS

    use MySQL::Benchmark::Logger qw( log );

=head1 DESCRIPTION

This module provides one single add-on method to be imported by any class
willing to implement C<log()>.

=head1 METHODS

=head2 log

=cut

sub log {
    my ( $self, @messages ) = @_;
    return if $$self{options}{quiet};

    my $debug_info = '';
    if ( $$self{options}{debug} ) {
        my ( $package, $filename, $line, $subroutine ) = caller 1;
        $package = ( split qr{::}, $package )[-1];
        $debug_info
            = $package . '['
            . $$ . ']: ' . '('
            . $subroutine . ':'
            . $line . ') ';
    }

    my ( $s, $ms ) = Time::HiRes::gettimeofday;
    my ( $sec, $min, $hour, $mday, $mon, $year ) = localtime($s);
    $year += 1900;
    print STDERR map {
        sprintf( '%04d-%02d-%02d %02d:%02d:%02d.%06.0f ',
            $year, $mon, $mday, $hour, $min, $sec, $ms )
            . $debug_info . $_ . "\n"
    } @messages;

    return;
}

=head1 AUTHOR

Luis Motta Campos, C<< <lmc at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Luis Motta Campos.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; version 2 dated June, 1991 or at your option any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree; if
not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.

=cut

1;    # End of MySQL::Benchmark::Logger


lib/MySQL/Benchmark/IPC/Client.pm
package MySQL::Benchmark::IPC::Client;

use 5.006;
use strict;
use warnings FATAL => 'all';

use IO::Socket;

=head1 NAME

MySQL::Benchmark::IPC::Client - An IPC Client for MySQL::Benchmark communication
between Controller and Workers.

=head1 SYNOPSYS

    use MySQL::Benchmark::IPC::Client;
    my $ipc_client = MySQL::Benchmark::IPC::Client->new;
    $ipc_client->send( 'some message' );

=head1 DESCRIPTION

This module provides a simple IPC client for message passing between the
Controller and the Workers.

=head1 METHODS

=head2 new 

Constructor.

=cut

sub new {
    my ( $class, %self ) = @_;

    die 'Cannot communicate without a named pipe!' unless $self{socket_file};
    $self{timeout} ||= 5;

    my $self = bless \%self, $class;

    $self->initialise_socket;
    return $self;
}

=head2 initialise_socket 

Initialise the client-side communication socket.

Called automatically by L</new()>.

=cut

sub initialise_socket {
    my ($self) = @_;
    $$self{SOCKET} = IO::Socket::UNIX->new(
        Peer    => $$self{socket_file},
        Type    => SOCK_DGRAM,
        Timeout => $$self{timeout}
    );
    die qq{Cannot initialise IPC communication socket: $!.}
        unless defined $$self{SOCKET};
}

=head2 send

Send a message. Text only. Serialisation is a must. Watchout for the message
size, you're limited by your system's maximum datagram packet size.

=cut

sub send {
    my ( $self, $message ) = @_;
    $$self{SOCKET}->send($message);
}

=head2 DESTROY

Destructor. Ensures we shut down the socket before garbage collection.

=cut 

sub DESTROY {
    my ($self) = @_;
    if ( defined $$self{SOCKET} ) { $$self{SOCKET}->shutdown; }
    delete $$self{SOCKET};
}

=head1 AUTHOR

Luis Motta Campos, C<< <lmc at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Luis Motta Campos.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; version 2 dated June, 1991 or at your option any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree; if
not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.

=cut

1;    # End of MySQL::Benchmark::IPC::Server


lib/MySQL/Benchmark/IPC/Server.pm
package MySQL::Benchmark::IPC::Server;

use 5.006;
use strict;
use warnings FATAL => 'all';

use IO::Socket;
use File::Temp ();

=head1 NAME

MySQL::Benchmark::IPC::Server – a UDP listener for Benchmark results IPC.

=head1 VERSION

=head1 SYNOPSYS

=head1 DESCRIPTION

=head1 METHODS

=head2 new

Constructor.

Takes an optional "port" argument. If unset, port defaults to 2020.
Takes an optional "max_lenght" argument. If unset, defaults to 4096.

=cut

sub new {
    my ( $class, %self ) = @_;

    # Defaults
    $self{max_length} = 4096;

    my $self = bless \%self, $class;

    $self->initialise_socket;
    return $self;
}

=head2 initialise_socket

Opens a socket to listen on.

=cut

sub initialise_socket {
    my ($self) = @_;
    $$self{socket_file} ||= File::Temp::mktemp( '/tmp/mysql-benchmark-socket.XXXXXX' );
    unlink $$self{socket_file} if -e $$self{socket_file};
    my $sock = IO::Socket::UNIX->new(
        Local => $$self{socket_file},
        Type  => SOCK_DGRAM,
    );
    die qq{Cannot create socket: $!.} unless $sock;
    $$self{SOCKET} = $sock;
}

=head2 socket_file

Returns the filename used as socket for this server.

=cut

sub socket_file { my ($self) = @_; return $$self{socket_file}; }

=head2 receive

Attempts to receive a message.

=cut

sub receive {
    my ($self) = @_;
    my $message;
    $$self{SOCKET}->recv( $message, $$self{max_length} );
    return $message;
}

=head1 AUTHOR

Luis Motta Campos, C<< <lmc at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Luis Motta Campos.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; version 2 dated June, 1991 or at your option any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree; if
not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.

=cut

1;    # End of MySQL::Benchmark::IPC::Server


lib/MySQL/Benchmark.pm
package MySQL::Benchmark;

use 5.006;
use strict;
use warnings FATAL => 'all';

use Getopt::Long;
use MySQL::Benchmark::IPC::Server;
use MySQL::Benchmark::Worker;
use MySQL::Benchmark::Query;
use MySQL::Benchmark::Logger qw(log);
use Time::HiRes;
use POSIX qw(:sys_wait_h strftime ceil);
use YAML::XS ();
use Storable ();
use List::Util();
use Data::Dumper();

=head1 NAME

MySQL::Benchmark - Custom MySQL Benchmarks made easy.

=head1 VERSION

Version 1.04

=cut

our $VERSION = '1.04';

=head1 SYNOPSIS

=head1 DESCRIPTION

This module tries to address a niche in MySQL benchmarking: there are little to
no good parallel tools to benchmark MySQL databases that are efficient and at
the same time allow customization.

=head1 METHODS

=head2 new

=cut

sub new {
    my ($class) = @_;
    my $self = bless {}, $class;

    $self->evaluate_command_line_options;
    $self->load_queries_file;
    $self->normalise_query_weights;
    $self->initialise_ipc;
    $self->fork_workers;
    $self->initialise_signal_handlers;
    $self->supervision_loop;
    $self->tear_down_ipc;
    $self->output_results;
}

=head2 evaluate_command_line_options

=cut

sub evaluate_command_line_options {
    my ($self) = @_;

    # Default configuration
    # FIXME: define workers in function of available processor cores?
    my $options = {
        workers              => 1,
        runtime              => 10,
        mysql                => { defaults_file => "$ENV{HOME}/.my.cnf" },
        flush_interval       => 3,
        max_query_array_size => 10,
        report               => 0,
    };

    my $version_handler = sub {
            use FindBin qw($Bin);
        print STDERR qq{$Bin running version $MySQL::Benchmark::VERSION.\n};
        exit;
    };

    my $result = GetOptions(
        'version'   => $version_handler,
        'debug'     => \$$options{debug},
        'quiet'     => \$$options{quiet},
        'verbose'   => \$$options{verbose},
        'report'    => \$$options{report},
        'csv'       => \$$options{csv},
        'outfile=s' => \$$options{outfile},

        'queries=s'              => \$$options{queries},
        'workers=i'              => \$$options{workers},
        'max-query-array-size=i' => \$$options{max_query_array_size},
        'runtime=i'              => \$$options{runtime},
        'flush-interval=i'       => \$$options{flush_interval},

        'dbhost=s'     => \$$options{mysql}{host},
        'dbschema=s'   => \$$options{mysql}{schema},
        'dbuser=s'     => \$$options{mysql}{user},
        'dbpassword=s' => \$$options{mysql}{password},
        'dbdefaults=s' => \$$options{mysql}{defaults_file},
    );

    $$self{options} = $options;
}

=head2 load_queries_file

=cut

sub load_queries_file {
    my ($self) = @_;

    unless ( -f $$self{options}{queries} ) {
        $self->log(qq(File "$$self{options}{queries}" doesn't exist.));
        exit;
    }

    $$self{queries} = [ map { MySQL::Benchmark::Query->new($_) }
            @{ YAML::XS::LoadFile( $$self{options}{queries} ) } ];

    if ( $$self{options}{debug} ) {
        $self->log( q(Loaded )
                . ( $$self{queries} ? scalar @{ $$self{queries} } : 0 )
                . qq( queries from file "$$self{options}{queries}".) );
    }

    return;
}

=head2 normalise_query_weights

Assigns proper weights for every query loaded from the Query Descriptor File.

=cut

sub normalise_query_weights {
    my ($self) = @_;

    # Calculate Total Weight from the Query Definition File
    my $total_weight = 0;
    foreach my $query ( @{ $$self{queries} } ) {
        $total_weight += $query->weight;
    }
    $self->log("Total Query Weight is $total_weight.")
        if $$self{options}{debug};

    # Normalise the weight of all queries to the requested array size
    my @new_queries = ();
    foreach my $query ( @{ $$self{queries} } ) {
        my $new_weight
            = ceil( $$self{options}{max_query_array_size}
                * $query->weight
                / $total_weight );
        $new_weight = $new_weight > 1 ? $new_weight : 1;
        $query->weight($new_weight);
        $self->log(
            qq(Query @{[$query->id]} has normalised weight of @{[$query->weight]}.)
        ) if $$self{options}{debug};

        # Reflect the change on the new Query Array.
        push @new_queries, $query for 1 .. $new_weight;
    }

    $$self{queries} = [@new_queries];

    return;
}

=head2 initialise_ipc

=cut 

sub initialise_ipc {
    my ($self) = @_;
    my $server = eval { MySQL::Benchmark::IPC::Server->new };
    if ($@) {
        $self->log(qq{Cannot initalise IPC communications: $@});
        exit;
    }
    $$self{ipc_server} = $server;
}

=head2 ipc_server_socket

Accessor. Returns the socket currently in use by the underlying
MySQL::Benchmark::IPC::Server.

=cut

sub ipc_server_socket {
    my ($self) = @_;
    return $$self{ipc_server}->socket_file;
}

=head2 fork_workers

=cut 

sub fork_workers {
    my ($self) = @_;
    my $workers;
    $$self{worker_pids} = [];
    while ( ++$workers <= $$self{options}{workers} ) {
        push @{ $$self{worker_pids} },
            MySQL::Benchmark::Worker->new(
            mysql          => $$self{options}{mysql},
            queries        => [ List::Util::shuffle( @{ $$self{queries} } ) ],
            flush_interval => $$self{options}{flush_interval},
            socket_file    => $self->ipc_server_socket,
            options        => {
                debug   => $$self{options}{debug},
                verbose => $$self{options}{verbose},
                quiet   => $$self{options}{quiet},
            },
            );
    }
    $self->log("Forked workers: @{ $$self{worker_pids} }.")
        if $$self{options}{debug};
}

=head2 tear_down_workers

=cut

sub tear_down_workers {
    my ($self) = @_;
    my $stopped_count;
    if ( $$self{options}{debug} ) {
        local $" = ', ';
        $self->log("Signalling workers: @{ $$self{worker_pids} }.");
    }
    foreach my $worker ( @{ $$self{worker_pids} } ) {
        if ( my $count = kill 'TERM', $worker ) {
            $stopped_count += $count;
        }
    }
    $self->log("Signalled $stopped_count workers.") if $$self{options}{debug};
}

=head2 stop_benchmark

=cut 

sub stop_benchmark {
    my ($self) = @_;
    return if $self->is_stopped;
    $self->log('Stopping benchmark') if $$self{options}{debug};
    $$self{STOP} = 1;
    $self->tear_down_workers;
}

=head2 is_stopped

=cut

sub is_stopped { ${ $_[0] }{STOP} }

=head2 handle_sigchild

=cut

sub handle_sigchild {
    my ($self) = @_;
    while ( ( my $kid = waitpid( -1, WNOHANG ) ) > 0 ) {
        @{ $$self{worker_pids} }
            = grep { $_ != $kid } @{ $$self{worker_pids} };
    }
}

=head2 handle_sigterm

=cut

sub handle_sigterm {
    my ($self) = @_;
    $self->log('    Interrupted by user. Stopping benchmark.');
    $self->stop_benchmark;
}

=head2 initialise_signal_handlers

=cut

sub initialise_signal_handlers {
    my ($self) = @_;
    $SIG{CHLD} = sub { $self->handle_sigchild };
    $SIG{TERM} = $SIG{INT} = sub { $self->handle_sigterm };
}

=head2 is_time_to_stop

=cut

sub is_time_to_stop {
    my ($self) = @_;

    if ( $self->is_stopped ) {
        return 1;
    }
    else {
        return Time::HiRes::tv_interval( $$self{start_time} )
            - $$self{options}{runtime} > 0;
    }
}

=head2 receive_message

=cut

sub receive_message {
    my ($self) = @_;
    my $frozen = $$self{ipc_server}->receive;

    # Ignore malformed messages without further ado.
    my $message = eval { Storable::thaw($frozen) };
    return $message;
}

=head2 process_message

=cut 

sub process_message {
    my ( $self, $message ) = @_;
    return
        unless UNIVERSAL::isa( $message, 'HASH' )
        && UNIVERSAL::isa( $$message{statistics}, 'HASH' );

    # The idea here is to maintain global totals and per-minute partials.
    foreach my $query ( keys %{ $$message{statistics} } ) {

        # Global Total
        $$self{global_stats}{totals}{runs}
            += $$message{statistics}{$query}{runs};
        $$self{global_stats}{totals}{run_time}
            += $$message{statistics}{$query}{run_time};
        $$self{global_stats}{totals}{bytes_sent}
            += $$message{statistics}{$query}{session}{bytes_sent};
        $$self{global_stats}{totals}{bytes_received}
            += $$message{statistics}{$query}{session}{bytes_received};

        # Per Query Totals
        $$self{global_stats}{per_query}{$query}{runs}
            += $$message{statistics}{$query}{runs};
        $$self{global_stats}{per_query}{$query}{run_time}
            += $$message{statistics}{$query}{run_time};
        $$self{global_stats}{per_query}{$query}{bytes_sent}
            += $$message{statistics}{$query}{session}{bytes_sent};
        $$self{global_stats}{per_query}{$query}{bytes_received}
            += $$message{statistics}{$query}{session}{bytes_received};
    }

}

=head2 supervision_loop

=cut

sub supervision_loop {
    my ($self) = @_;
    $$self{start_time} = [Time::HiRes::gettimeofday];

SUPERVISION_LOOP:
    while ( scalar @{ $$self{worker_pids} } > 0 ) {

        my $message = $self->receive_message;
        $self->process_message($message);

    }
    continue {
        if ( $self->is_time_to_stop && !$self->is_stopped ) {
            $self->stop_benchmark;
        }
    }

    $$self{end_time} = [Time::HiRes::gettimeofday];

}

=head2 tear_down_ipc

=cut

sub tear_down_ipc { }

=head2 __format_times

=cut

sub __format_times {
    my ($self)     = @_;
    my @start_time = @{ $$self{start_time} };
    my @end_time   = @{ $$self{end_time} };
    my $start_time
        = strftime( '%Y-%m-%d %H:%M:%S.', localtime( $start_time[0] ) )
        . $start_time[1];
    my $end_time = strftime( '%Y-%m-%d %H:%M:%S.', localtime( $end_time[0] ) )
        . $end_time[1];

    my $real_clock_run_time = Time::HiRes::tv_interval( $$self{start_time} );

    @{ $$self{formatted} }{qw(start end real)}
        = ( $start_time, $end_time, $real_clock_run_time );
    return;
}

=head2 output_results

=cut

sub output_results {
    my ($self) = @_;

    $self->__format_times;

    # Either open the right file or redirect the output
    my $outfile;
    if ( $$self{options}{outfile} ) {
        open $outfile, '>', $$self{options}{outfile} or die $!;
    }
    else { $outfile = \*STDOUT; }

    if ( $$self{options}{report} ) {
        print $outfile $self->output_report;
    }

    if ( $$self{options}{csv} ) {
        print $outfile $self->output_csv;
    }

    if ( $$self{options}{outfile} ) { close $outfile or die $!; }
}

=head2 output_csv

=cut

sub output_csv {
    my ($self) = @_;
    my $result = join( ',',
        'Start Time',
        'End Time',
        'Real Elapsed Time',
        'Workers',
        'Run Time',
        'Runs',
        'Bytes Sent',
        'Bytes Received' )
        . "\n";
    $result .= join( ',',
        "'$$self{formatted}{start}'",
        "'$$self{formatted}{end}'",
        $$self{formatted}{real},
        $$self{options}{workers},
        @{ $$self{global_stats}{totals} }
            {qw( run_time runs bytes_sent bytes_received )} )
        . "\n";
    $result .= "\n";
    $result .= join(
        "\n",
        join( ',',
            'Query ID',
            'Run Time',
            'Runs',
            'Bytes Sent',
            'Bytes Received' ),
        map {
            join( ',',
                "'$_'",
                @{ $$self{global_stats}{per_query}{$_} }
                    {qw( run_time runs bytes_sent bytes_received )} )
            }
            keys %{ $$self{global_stats}{per_query} }
    ) . "\n";
    return $result;
}

=head2 output_report

=cut

sub output_report {
    my ($self) = @_;

    my $result
        = qq(\nBenchmark Complete.)
        . qq(\n\n\tStart Time: $$self{formatted}{start})
        . qq(\n\tEnd Time: $$self{formatted}{end})
        . qq(\n\tReal Clock Elapsed Time: $$self{formatted}{real} seconds.)
        . qq(\n\tUsed $$self{options}{workers} worker processes.\n)
        . qq(\n\tPER QUERY STATISTICS:)
        . join(
        "\n",
        map {
                  qq(\n\t\tQuery ID: $_)
                . qq(\n\t\t\tRun Time: $$self{global_stats}{per_query}{$_}{run_time})
                . qq(\n\t\t\tRuns: $$self{global_stats}{per_query}{$_}{runs})
                . qq(\n\t\t\tBytes Sent: $$self{global_stats}{per_query}{$_}{bytes_sent})
                . qq(\n\t\t\tBytes Received: $$self{global_stats}{per_query}{$_}{bytes_received})
            }
            keys %{ $$self{global_stats}{per_query} }
        )
        . qq(\n\n\tGLOBAL STATISTICS:)
        . qq(\n\t\tRun Time: $$self{global_stats}{totals}{run_time})
        . qq(\n\t\tRuns: $$self{global_stats}{totals}{runs})
        . qq(\n\t\tBytes Sent: $$self{global_stats}{totals}{bytes_sent})
        . qq(\n\t\tBytes Received: $$self{global_stats}{totals}{bytes_received})
        . qq{\n\n\n};

    return $result;
}

=head1 AUTHOR

Luis Motta Campos, C << <lmc at cpan.org> >>

=head1 BUGS

The web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=mysql-benchmark>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MySQL::Benchmark


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=mysql-benchmark>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/mysql-benchmark>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/mysql-benchmark>

=item * Search CPAN

L<http://search.cpan.org/dist/mysql-benchmark/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Luis Motta Campos.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; version 2 dated June, 1991 or at your option any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree; if
not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.

=cut

1;    # End of MySQL::Benchmark
